pkg/domain/blueprintSpec.go

```diff
@@ -218,7 +223,35 @@ func (spec *BlueprintSpec) calculateEffectiveDogu(dogu Dogu) (Dogu, error) {
 	return effectiveDogu, nil
 }
 
+// MarkInvalid is used to mark the blueprint as invalid after dynamically validating it.
 func (spec *BlueprintSpec) MarkInvalid(err error) {
 	spec.Status = StatusPhaseInvalid
 	spec.Events = append(spec.Events, BlueprintSpecInvalidEvent{ValidationError: err})
 }
+
+// DetermineStateDiff creates the StateDiff between the blueprint and the actual state of the ecosystem.
+// if sth. is not in the lists of installed things, it is considered not installed.
+// installedDogus are a map in the form of simpleDoguName->*DoguInstallation. There should be no nil values.
+// The StateDiff is an 'as is' representation, therefore no error is thrown, e.g. if dogu namespaces are different and namespace changes are not allowed.
+// If there are not allowed actions should be considered at the start of the execution of the blueprint.
+// returns an error if the BlueprintSpec is not in the necessary state to determine the stateDiff.
+func (spec *BlueprintSpec) DetermineStateDiff(installedDogus map[string]*ecosystem.DoguInstallation) error {
+	switch spec.Status {
+	case StatusPhaseNew:
+		return fmt.Errorf("cannot determine state diff in status phase %q", spec.Status)
+	case StatusPhaseStaticallyValidated:
+		return fmt.Errorf("cannot determine state diff in status phase %q", spec.Status)
+	case StatusPhaseValidated:
+		return fmt.Errorf("cannot determine state diff in status phase %q", spec.Status)
+	case StatusPhaseEffectiveBlueprintGenerated: // this is the state, the blueprint spec should be
+	default:
+		return nil // do not re-determine the state diff from status StatusPhaseStateDiffDetermined and above
+	}
+	spec.StateDiff = StateDiff{
+		DoguDiffs: determineDoguDiffs(spec.EffectiveBlueprint.Dogus, installedDogus),
+		// there will be more diffs, e.g. for components and registry keys
+	}
+	spec.Status = StatusPhaseStateDiffDetermined
+	spec.Events = append(spec.Events, StateDiffDeterminedEvent{StateDiff: spec.StateDiff})
```

Comment:
An additional generalized event would be useful (like: X Diffs found (Y to install, Z to upgrade, A to delete))

Reply:
Will be part of the next PR where we're currently overhauling events.